[AUTH-133] Add pod security admission control docs (#47647)
openshift/openshift-docs: modules/security-context-constraints-psa-opting.adoc

// Module included in the following assemblies:
//
// * authentication/managing-security-context-constraints.adoc

:_content-type: PROCEDURE
[id="security-context-constraints-psa-opting_{context}"]
= Disabling pod security admission synchronization

> You can disable automatic pod security admission synchronization for a namespace.

Reviewer: What about the namespaces that are disabled by default? And those that cannot ever be synchronized?

Author: It's a discrete module that describes the task of disabling the feature in namespaces where it is active. 

I can do one of two things for your first case:
* Change this to a more generic "setting synchronization" module, which could feel clunky in the end. 
* Add an "Enabling pod security admission synchronization" module, which would be quite similar to this one. 

The second case should be covered by `modules/security-context-constraints-psa-synchronization.adoc`, but we can nod to namespaces that cannot ever be synchronized here, too. 

Thoughts?

Author: Trying out the second approach. See `modules/security-context-constraints-psa-opting-in.adoc`. @stlaz

Author: As I'm likely not going to be owning this long-term, I'm game for that. 👍

.Procedure

> * For each namespace for which you want to disable sychronization, change the value of the `security.openshift.io/scc.podSecurityLabelSync` label to `false`. 

Reviewer: I'd rather use "set" instead of "change" because the label is unlikely to be set.

Author: ```suggestion
* For each namespace for which you want to disable sychronization, set the value of the `security.openshift.io/scc.podSecurityLabelSync` label to `false`. 
```

// Module included in the following assemblies:
//
// * authentication/managing-security-context-constraints.adoc

:_content-type: PROCEDURE
[id="security-context-constraints-psa-opting_{context}"]
= Controlling pod security admission synchronization

> You can enable or disable automatic pod security admission synchronization for most namespaces.

Reviewer 2: We need to make clear in the docs that is NOT possible to enable the synchronization for the ns which are part of the OCP payload: 

https://github.com/openshift/cluster-policy-controller/blob/c7201edacb6e4ac1f8a1624b152911f3c404dbc7/pkg/psalabelsyncer/podsecurity_label_sync_controller.go#L42-L111

The docs should contain the list or a link for the implementation ( release tag 4.11 ) : https://github.com/openshift/cluster-policy-controller/blob/release-4.11/pkg/psalabelsyncer/podsecurity_label_sync_controller.go#L42-L111

Reviewer: I wanted to specifically leave this vague in case it changes

[NOTE]
====

> By default, namespaces that have the prefix 'openshift-' have pod security admission synchronization disabled. 

Reviewer 2: @maxwelldb any namespace prefixed with "openshift" has the sync disabled by default
Therefore, for that namespace which are part of the OCP payload has no way of we make the label syncer work at all. These ones are mainly prefixed with "openshift" but we cannot say that is created by the installer because we created "openshift-operators" for example and this one is NOT part of the OCP payload then we need to clarify in IMO the best way is to link the list of those namespaces which are part of the OCP payload ( from the code ) and will never have the sync working at all.

> By default, namespaces that have the prefix 'openshift-' have pod security admission synchronization disabled. 

Reviewer: > So is it the case that, as a user, if I create a namespace called openshift-user-cool-namespace, sync is disabled by default?

yes

> These ones are mainly prefixed with "openshift" but we cannot say that is created by the installer because we created "openshift-operators"

"openshift-operators" is the only exception, you can therefore call it that but this one exception should be specifically called out.

I don't want a specific list of namespaces because I am not convinced about its stabiity.

> Namespaces that the installer creates have pod security admission synchronization disabled permanently.

Reviewer 2: @stlaz it is not true. See that `openshift-operators` is created by the installer and allows the PSA because it is not a OCP playload namespace:

```suggestion
Namespaces that are defined as OCP payloads have pod security admission synchronization disabled permanently. You can check the list of those namespaces [here](https://github.com/openshift/cluster-policy-controller/blob/release-4.11/pkg/psalabelsyncer/podsecurity_label_sync_controller.go#L42-L111).
```

Reviewer: I believe the docs should never have links to the code.

Also, yes, the `openshift-operators` is unfortunately special, I forgot about it, again :|

Author: ```suggestion
Namespaces that are defined as part of the cluster payload have pod security admission synchronization disabled permanently. For a reference for these namespaces, see "List of payload namespaces".
```

Added document `modules/security-context-constraints-psa-payload-namespaces.adoc`. We should not be linking to code whenever possible.

Reviewer 2: Hi @maxwelldb: 
We can add a doc with the list of the namespace however, we cannot say

` Namespaces that the installer creates ` Not all namespaces prefixed with "openshift" created by the installer are part of the OCP Playload and have the limitation that will never have the label syncer working there.

Reviewer: > Namespaces that the installer creates have pod security admission synchronization disabled permanently.

This was the better version, nobody says that every namespace prefixed with "openshift-" was created by the installer.

Possibly add `(with the exception of openshift-operators namespace)` somewhere

Author: ```suggestion
Namespaces that the installer creates have pod security admission disabled permanently. These namespaces include:

* All namespaces that are prefixed with `openshift-`, except for `openshift-operators`
* `default``
* `kube-node-lease`
* `kube-system`      
* `kube-public`      
* `openshift`
```

// Module included in the following assemblies:
//
// * authentication/managing-security-context-constraints.adoc

:_content-type: PROCEDURE
[id="security-context-constraints-psa-opting_{context}"]
= Controlling pod security admission synchronization

You can enable or disable automatic pod security admission synchronization for most namespaces.

[NOTE]
====

> By default, user-created namespaces that have the prefix 'openshift-' have pod security admission synchronization disabled. 
> 
> Namespaces that the installer creates have pod security admission disabled permanently. These namespaces include:

Reviewer: ```suggestion
By default, user-created namespaces that have the prefix 'openshift-' have pod security admission label synchronization disabled. 

Namespaces that the installer creates have pod security admission label synchronization disabled permanently. These namespaces include:
```

* All namespaces that are prefixed with `openshift-`, except for `openshift-operators`
* `default``
* `kube-node-lease`
* `kube-system`      
* `kube-public`      
* `openshift`
====

.Procedure

* For each namespace that you want to configure, set a value for the `security.openshift.io/scc.podSecurityLabelSync` label: 

> ** To disable pod security admission sychronization in a namespace, set the value of the `security.openshift.io/scc.podSecurityLabelSync` label to `false`.  

Reviewer: ```suggestion
** To disable pod security admission label sychronization in a namespace, set the value of the `security.openshift.io/scc.podSecurityLabelSync` label to `false`.  
```

+
For example, from a command line, enter:
+
[source,terminal]
----
$ oc label namespace my-namespace security.openshift.io/scc.podSecurityLabelSync=false
----

> ** To enable pod security admission sychronization in a namespace, set the value of the `security.openshift.io/scc.podSecurityLabelSync` label to `true`. 

Reviewer: ```suggestion
** To enable pod security admission label sychronization in a namespace, set the value of the `security.openshift.io/scc.podSecurityLabelSync` label to `true`. 
```
